File: ctci/one_away.py
# from ctci

def one_away(s, t):
    n, m = len(s), len(t)
    if n - m > 1 or n - m < -1: return False
    i, j = 0, 0
    replaced, inserted = False, False
    while i < n and j < m:
        if s[i] == t[j]:
            i += 1
            j += 1
        elif n == m and not replaced:
            i += 1
            j += 1
            replaced = True
        elif n != m and not inserted:
            if n > m: i += 1
            elif m > n: j += 1
            inserted = True
        else: return False
    return True

assert one_away('pale', 'ple')
assert one_away('pales', 'pale')
assert one_away('pale', 'bale')
assert not one_away('pale', 'bake')

File: ctci/OneAway.java
import java.lang.Math;

public class OneAway {
	public boolean oneAway(String s0, String s1) {
		if (s0.length() != s1.length() && Math.abs(s0.length() - s1.length()) != 1)
			return false;

		// check for replacement
		if (s0.length() == s1.length()) {
			int diff = 0;

			for (int i = 0; i < s0.length(); i++) {
				if (diff > 1) return false;
				if (s0.charAt(i) != s1.charAt(i)) diff++;
			}
			return diff == 1;
		}
		else { // check for insertion/deletion
			boolean foundMismatch = false;
			int l = 0; 
			int s = 0;

			String s_long = s0.length() < s1.length() ? s1 : s0;
			String s_short = s0.length() < s1.length() ? s0 : s1;

			for (; s < s_short.length(); s++) {
				if (s_long.charAt(l) != s_short.charAt(s)) {
					if (foundMismatch) return false;
					foundMismatch = true;
					l++;
				}
				l++;
			}
			return true;
		}
	}

	public static void main(String[] args) {
		OneAway oa = new OneAway();

		// true
		System.out.println(oa.oneAway("pale", "ple"));
		System.out.println(oa.oneAway("ple", "pale"));
		System.out.println(oa.oneAway("pales", "pale"));
		System.out.println(oa.oneAway("pales", "paless"));
		System.out.println(oa.oneAway("hello", "helllo"));

		// false
		System.out.println(oa.oneAway("hello", "hello"));
		System.out.println(oa.oneAway("pale", "bake"));
		System.out.println(oa.oneAway("", ""));
	}
}